gh gist delete 38003

### Describe the bug

gh version 2.8.0 (2022-04-13)


### Steps to reproduce the behavior

see logs

### Expected vs actual behavior


### Logs

```
 $>gh gist delete 38003
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xdf6acf]

goroutine 1 [running]:
github.com/cli/cli/v2/pkg/cmd/gist/delete.deleteRun(0xc00061be90, 0x0, 0x0)
        /home/runner/work/cli/cli/pkg/cmd/gist/delete/delete.go:82 +0x1cf
github.com/cli/cli/v2/pkg/cmd/gist/delete.NewCmdDelete.func1(0xc0001a7400, 0xc000463dd0, 0x1, 0x1, 0x0, 0x0)
        /home/runner/work/cli/cli/pkg/cmd/gist/delete/delete.go:40 +0x8c
github.com/spf13/Paste the activity from your command line. Redact if needed.
cobra.(*Command).execute(0xc0001a7400, 0xc000463db0, 0x1, 0x1, 0xc0001a7400, 0xc000463db0)
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:856 +0x472
github.com/spf13/cobra.(*Command).ExecuteC(0xc0007df680, 0xc0007222c0, 0x3, 0x3)
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:974 +0x375
main.mainRun(0xc00007c058)
        /home/runner/work/cli/cli/cmd/gh/main.go:204 +0xaf4
main.main()
        /home/runner/work/cli/cli/cmd/gh/main.go:47 +0x25
```

--- a/pkg/cmd/gist/delete/delete.go
+++ b/pkg/cmd/gist/delete/delete.go
@@ -2,6 +2,7 @@ package delete
 
 import (
 	"errors"
+	"fmt"
 	"net/http"
 	"strings"
 
@@ -58,8 +59,6 @@ func deleteRun(opts *DeleteOptions) error {
 		return err
 	}
 
-	apiClient := api.NewClientFromHTTP(client)
-
 	cfg, err := opts.Config()
 	if err != nil {
 		return err
@@ -70,21 +69,11 @@ func deleteRun(opts *DeleteOptions) error {
 		return err
 	}
 
-	gist, err := shared.GetGist(client, host, gistID)
-	if err != nil {
-		return err
-	}
-	username, err := api.CurrentLoginName(apiClient, host)
-	if err != nil {
-		return err
-	}
-
-	if username != gist.Owner.Login {
-		return errors.New("you do not own this gist")
-	}
-
-	err = deleteGist(apiClient, host, gistID)
-	if err != nil {
+	apiClient := api.NewClientFromHTTP(client)
+	if err := deleteGist(apiClient, host, gistID); err != nil {
+		if errors.Is(err, shared.NotFoundErr) {
+			return fmt.Errorf("unable to delete gist %s: either the gist is not found or it is not owned by you", gistID)
+		}
 		return err
 	}
 
@@ -95,6 +84,10 @@ func deleteGist(apiClient *api.Client, hostname string, gistID string) error {
 	path := "gists/" + gistID
 	err := apiClient.REST(hostname, "DELETE", path, nil, nil)
 	if err != nil {
+		var httpErr api.HTTPError
+		if errors.As(err, &httpErr) && httpErr.StatusCode == 404 {
+			return shared.NotFoundErr
+		}
 		return err
 	}
 	return nil

--- a/pkg/cmd/gist/delete/delete_test.go
+++ b/pkg/cmd/gist/delete/delete_test.go
@@ -6,7 +6,6 @@ import (
 	"testing"
 
 	"github.com/cli/cli/v2/internal/config"
-	"github.com/cli/cli/v2/pkg/cmd/gist/shared"
 	"github.com/cli/cli/v2/pkg/cmdutil"
 	"github.com/cli/cli/v2/pkg/httpmock"
 	"github.com/cli/cli/v2/pkg/iostreams"
@@ -57,71 +56,44 @@ func TestNewCmdDelete(t *testing.T) {
 func Test_deleteRun(t *testing.T) {
 	tests := []struct {
 		name       string
-		opts       *DeleteOptions
-		gist       *shared.Gist
+		opts       DeleteOptions
 		httpStubs  func(*httpmock.Registry)
-		nontty     bool
 		wantErr    bool
+		wantStdout string
 		wantStderr string
-		wantParams map[string]interface{}
 	}{
 		{
-			name:    "no such gist",
-			wantErr: true,
-		}, {
-			name: "another user's gist",
-			gist: &shared.Gist{
-				ID: "1234",
-				Files: map[string]*shared.GistFile{
-					"cicada.txt": {
-						Filename: "cicada.txt",
-						Content:  "bwhiizzzbwhuiiizzzz",
-						Type:     "text/plain",
-					},
-				},
-				Owner: &shared.GistOwner{Login: "octocat2"},
-			},
-			wantErr:    true,
-			wantStderr: "you do not own this gist",
-		}, {
 			name: "successfully delete",
-			gist: &shared.Gist{
-				ID: "1234",
-				Files: map[string]*shared.GistFile{
-					"cicada.txt": {
-						Filename: "cicada.txt",
-						Content:  "bwhiizzzbwhuiiizzzz",
-						Type:     "text/plain",
-					},
-				},
-				Owner: &shared.GistOwner{Login: "octocat"},
+			opts: DeleteOptions{
+				Selector: "1234",
 			},
 			httpStubs: func(reg *httpmock.Registry) {
 				reg.Register(httpmock.REST("DELETE", "gists/1234"),
-					httpmock.StringResponse("{}"))
+					httpmock.StatusStringResponse(200, "{}"))
 			},
-			wantErr: false,
+			wantErr:    false,
+			wantStdout: "",
+			wantStderr: "",
+		},
+		{
+			name: "not found",
+			opts: DeleteOptions{
+				Selector: "1234",
+			},
+			httpStubs: func(reg *httpmock.Registry) {
+				reg.Register(httpmock.REST("DELETE", "gists/1234"),
+					httpmock.StatusStringResponse(404, "{}"))
+			},
+			wantErr:    true,
+			wantStdout: "",
+			wantStderr: "",
 		},
 	}
 
 	for _, tt := range tests {
 		reg := &httpmock.Registry{}
-		if tt.gist == nil {
-			reg.Register(httpmock.REST("GET", "gists/1234"),
-				httpmock.StatusStringResponse(404, "Not Found"))
-		} else {
-			reg.Register(httpmock.REST("GET", "gists/1234"),
-				httpmock.JSONResponse(tt.gist))
-			reg.Register(httpmock.GraphQL(`query UserCurrent\b`),
-				httpmock.StringResponse(`{"data":{"viewer":{"login":"octocat"}}}`))
-		}
-
 		if tt.httpStubs != nil {
 			tt.httpStubs(reg)
-		}
-
-		if tt.opts == nil {
-			tt.opts = &DeleteOptions{}
 		}
 
 		tt.opts.HttpClient = func() (*http.Client, error) {
@@ -130,14 +102,13 @@ func Test_deleteRun(t *testing.T) {
 		tt.opts.Config = func() (config.Config, error) {
 			return config.NewBlankConfig(), nil
 		}
-		io, _, _, _ := iostreams.Test()
-		io.SetStdoutTTY(!tt.nontty)
-		io.SetStdinTTY(!tt.nontty)
-		tt.opts.IO = io
-		tt.opts.Selector = "1234"
+		ios, _, _, _ := iostreams.Test()
+		ios.SetStdoutTTY(false)
+		ios.SetStdinTTY(false)
+		tt.opts.IO = ios
 
 		t.Run(tt.name, func(t *testing.T) {
-			err := deleteRun(tt.opts)
+			err := deleteRun(&tt.opts)
 			reg.Verify(t)
 			if tt.wantErr {
 				assert.Error(t, err)

